src/common/libflux/message.c

```diff
@@ -38,17 +39,50 @@
 
 #include "src/common/libutil/aux.h"
 #include "src/common/libutil/errno_safe.h"
+#include "src/common/libccan/ccan/list/list.h"
 
 #include "message.h"
 
 struct flux_msg {
-    zmsg_t *zmsg;
+    // optional route list, if FLUX_MSGFLAG_ROUTE
+    struct list_head routes;
+    int routes_len;     /* to avoid looping */
+
+    // optional topic frame, if FLUX_MSGFLAG_TOPIC
+    char *topic;
+
+    // optional payload frame, if FLUX_MSGFLAG_PAYLOAD
+    void *payload;
+    size_t payload_size;
+
+    // required proto frame
+    uint8_t magic;
+    uint8_t version;
```

Comment:
magic and version could probably be left out of the struct and just tacked on when the message is encoded/decoded?